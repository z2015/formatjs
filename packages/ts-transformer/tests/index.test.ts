import {join} from 'path';
import {transform, Opts, MessageDescriptor} from '../';
import * as ts from 'typescript';
import {readFile as readFileAsync} from 'fs';
import {promisify} from 'util';

const readFile = promisify(readFileAsync);

const FILES_TO_TESTS: Record<string, Partial<Opts>> = {
  additionalComponentNames: {
    additionalComponentNames: ['CustomMessage'],
  },
  defineMessages: {},
  extractFromFormatMessage: {
    extractFromFormatMessageCall: true,
  },
  extractFromFormatMessageStateless: {
    extractFromFormatMessageCall: true,
  },
  nested: {
    extractFromFormatMessageCall: true,
    overrideIdFn: (id, defaultMessage, description) => {
      return `HELLO.${id}.${defaultMessage!.length}.${typeof description}`;
    },
  },
  extractSourceLocation: {
    extractSourceLocation: true,
  },
  formatMessageCall: {},
  FormattedMessage: {},
  inline: {},
  overrideIdFn: {
    overrideIdFn: (id, defaultMessage, description) => {
      return `HELLO.${id}.${defaultMessage!.length}.${typeof description}`;
    },
    extractFromFormatMessageCall: true,
  },
  removeDefaultMessage: {
    removeDefaultMessage: true,
  },
  noImport: {
    overrideIdFn: '[sha1:contenthash:base64:6]',
    extractFromFormatMessageCall: true,
  },
  removeDescription: {},
};

const FIXTURES_DIR = join(__dirname, 'fixtures');

describe('emit asserts for', function () {
  const filenames = Object.keys(FILES_TO_TESTS);
  filenames.forEach(function (fn) {
    if (fn === 'extractSourceLocation') {
      it(`[special] ${fn}`, async function () {
        const output = await compile(
          join(FIXTURES_DIR, `${fn}.tsx`),
          FILES_TO_TESTS[fn]
        );
        // Check code output
        expect(output.code).toMatchSnapshot();
        expect(output.msgs).toHaveLength(1);
        expect(output.msgs[0]).toMatchSnapshot({
          defaultMessage: 'Hello World!',
          id: 'foo.bar.baz',
          start: 154,
          end: 222,
          file: expect.stringContaining('extractSourceLocation.tsx'),
        });
      });
    } else {
      it(fn, async function () {
        const output = await compile(
          join(FIXTURES_DIR, `${fn}.tsx`),
          FILES_TO_TESTS[fn]
        );
        expect(output).toMatchSnapshot();
      });
    }
  });
});

async function compile(filePath: string, options?: Partial<Opts>) {
  let msgs: MessageDescriptor[] = [];

  const input = await readFile(filePath, 'utf8');
  const output = ts.transpileModule(input, {
    compilerOptions: {
      experimentalDecorators: true,
      jsx: ts.JsxEmit.React,
      module: ts.ModuleKind.CommonJS,
      moduleResolution: ts.ModuleResolutionKind.NodeJs,
      noEmitOnError: false,
      noUnusedLocals: true,
      noUnusedParameters: true,
      stripInternal: true,
      declaration: false,
      baseUrl: __dirname,
      target: ts.ScriptTarget.ES2015,
      rootDir: __dirname,
      outDir: join(__dirname, 'output'),
    },
    fileName: filePath,
    reportDiagnostics: true,
    transformers: {
      before: [
        transform({
          overrideIdFn: '[hash:base64:10]',
          onMsgExtracted: (_, extractedMsgs) => {
            msgs = msgs.concat(extractedMsgs);
          },
          ...(options || {}),
        }),
      ],
    },
  });
  return {
    msgs,
    code: output.outputText,
  };
}
