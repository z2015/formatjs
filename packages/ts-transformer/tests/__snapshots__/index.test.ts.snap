// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emit asserts for [special] extractSourceLocation 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
class Foo extends react_1.Component {
    render() {
        return react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo.bar.baz', defaultMessage: 'Hello World!' });
    }
}
exports.default = Foo;
"
`;

exports[`emit asserts for [special] extractSourceLocation 2`] = `
Object {
  "defaultMessage": "Hello World!",
  "end": 222,
  "file": StringContaining "extractSourceLocation.tsx",
  "id": "foo.bar.baz",
  "start": 154,
}
`;

exports[`emit asserts for FormattedMessage 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo.bar.baz', defaultMessage: 'Hello World! {foo, number}', values: {
                foo: 1,
            } }));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World! {foo, number}",
      "description": "The default message.",
      "id": "foo.bar.baz",
    },
  ],
}
`;

exports[`emit asserts for additionalComponentNames 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
function CustomMessage() { }
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(CustomMessage, { id: 'greeting-world', defaultMessage: 'Hello World!' }));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "Greeting to the world",
      "id": "greeting-world",
    },
  ],
}
`;

exports[`emit asserts for defineMessages 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const msgs = react_intl_1.defineMessages({
    header: {
        id: \\"foo.bar.baz\\",
        defaultMessage: \\"Hello World!\\"
    },
    content: {
        id: \\"foo.bar.biff\\",
        defaultMessage: \\"Hello Nurse!\\"
    },
    kittens: {
        id: \\"app.home.kittens\\",
        defaultMessage: \\"{count, plural, =0 {\\\\uD83D\\\\uDE2D} one {# kitten} other {# kittens}}\\"
    },
    trailingWhitespace: {
        id: \\"trailing.ws\\",
        defaultMessage: \\"   Some whitespace   \\"
    },
    escaped: {
        id: \\"escaped.apostrophe\\",
        defaultMessage: \\"A quoted value ''{value}'\\"
    },
    quoted: {
        id: \\"escaped.apostrophe\\",
        defaultMessage: \\"What's going on\\"
    },
    newline: {
        id: \\"newline\\",
        defaultMessage: \\"this is     a message\\"
    }
});
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(\\"h1\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, Object.assign({}, msgs.header))),
            react_1.default.createElement(\\"p\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, Object.assign({}, msgs.content))),
            react_1.default.createElement(\\"p\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, Object.assign({}, msgs.kittens)))));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message",
      "id": "foo.bar.baz",
    },
    Object {
      "defaultMessage": "Hello Nurse!",
      "description": "Another message",
      "id": "foo.bar.biff",
    },
    Object {
      "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
      "description": "Counts kittens",
      "id": "app.home.kittens",
    },
    Object {
      "defaultMessage": "   Some whitespace   ",
      "description": "Whitespace",
      "id": "trailing.ws",
    },
    Object {
      "defaultMessage": "A quoted value ''{value}'",
      "description": "Escaped apostrophe",
      "id": "escaped.apostrophe",
    },
    Object {
      "defaultMessage": "What's going on",
      "description": "Escaped apostrophe",
      "id": "escaped.apostrophe",
    },
    Object {
      "defaultMessage": "this is     a message",
      "description": "this is     a     description",
      "id": "newline",
    },
  ],
}
`;

exports[`emit asserts for extractFromFormatMessage 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const objectPointer = {
    id: 'foo.bar.invalid',
    defaultMessage: 'This cannot be extracted',
    description: 'the plugin only supports inline objects',
};
class Foo extends react_1.Component {
    render() {
        const { intl } = this.props;
        const msgs = {
            baz: this.props.intl.formatMessage({
                id: \\"foo.bar.baz\\",
                defaultMessage: \\"Hello World!\\"
            }),
            biff: intl.formatMessage({
                id: \\"foo.bar.biff\\",
                defaultMessage: \\"Hello Nurse!\\"
            }),
            invalid: this.props.intl.formatMessage(objectPointer),
            invalid2: this.props.intl.formatMessage({
                id,
                defaultMessage,
                description: 'asd',
            }),
        };
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(\\"h1\\", null, msgs.header),
            react_1.default.createElement(\\"p\\", null, msgs.content),
            react_1.default.createElement(\\"input\\", { placeholder: intl.formatMessage({ id: \\"3in0OrMYu4\\", defaultMessage: \\"inline\\" }) }),
            react_1.default.createElement(\\"span\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo', defaultMessage: 'bar' }))));
    }
}
exports.default = react_intl_1.injectIntl(Foo);
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message",
      "id": "foo.bar.baz",
    },
    Object {
      "defaultMessage": "Hello Nurse!",
      "description": "Another message",
      "id": "foo.bar.biff",
    },
    Object {
      "defaultMessage": "inline",
      "id": "3in0OrMYu4",
    },
    Object {
      "defaultMessage": "bar",
      "description": "baz",
      "id": "foo",
    },
  ],
}
`;

exports[`emit asserts for extractFromFormatMessageStateless 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_intl_1 = require(\\"react-intl\\");
const react_1 = require(\\"react\\");
function myFunction(param1, { formatMessage, formatDate }) {
    return (formatMessage({
        id: \\"inline1\\",
        defaultMessage: \\"Hello params!\\"
    }) + formatDate(new Date()));
}
const child = myFunction(filterable, intl);
function SFC() {
    const { formatMessage } = react_intl_1.useIntl();
    return formatMessage({
        id: \\"hook\\",
        defaultMessage: \\"hook\\"
    });
}
const Foo = ({ intl: { formatMessage } }) => {
    const msgs = {
        qux: formatMessage({
            id: \\"foo.bar.quux\\",
            defaultMessage: \\"Hello Stateless!\\"
        }),
    };
    return (react_1.default.createElement(\\"div\\", null,
        react_1.default.createElement(\\"h1\\", null, msgs.header),
        react_1.default.createElement(\\"p\\", null, msgs.content),
        react_1.default.createElement(\\"span\\", null,
            react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo', defaultMessage: 'bar' }))));
};
exports.default = react_intl_1.injectIntl(Foo);
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello params!",
      "description": "A stateless message",
      "id": "inline1",
    },
    Object {
      "defaultMessage": "hook",
      "description": "hook",
      "id": "hook",
    },
    Object {
      "defaultMessage": "Hello Stateless!",
      "description": "A stateless message",
      "id": "foo.bar.quux",
    },
    Object {
      "defaultMessage": "bar",
      "description": "baz",
      "id": "foo",
    },
  ],
}
`;

exports[`emit asserts for formatMessageCall 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const objectPointer = {
    id: 'foo.bar.invalid',
    defaultMessage: 'This cannot be extracted',
    description: 'the plugin only supports inline objects',
};
class Foo extends react_1.Component {
    render() {
        const msgs = {
            baz: this.props.intl.formatMessage({
                id: 'foo.bar.baz',
                defaultMessage: 'Hello World!',
                description: 'The default message',
            }),
            biff: this.props.intl.formatMessage({
                id: 'foo.bar.biff',
                defaultMessage: 'Hello Nurse!',
                description: 'Another message',
            }),
            invalid: this.props.intl.formatMessage(objectPointer),
        };
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(\\"h1\\", null, msgs.header),
            react_1.default.createElement(\\"p\\", null, msgs.content),
            react_1.default.createElement(\\"input\\", { placeholder: intl.formatMessage({ defaultMessage: 'inline' }) }),
            react_1.default.createElement(\\"span\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo', defaultMessage: 'bar' }))));
    }
}
exports.default = react_intl_1.injectIntl(Foo);
",
  "msgs": Array [
    Object {
      "defaultMessage": "bar",
      "description": "baz",
      "id": "foo",
    },
  ],
}
`;

exports[`emit asserts for inline 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo.bar.baz', defaultMessage: 'Hello World!' }),
            react_intl_1.defineMessage({
                id: \\"header\\",
                defaultMessage: \\"Hello World!\\"
            }),
            react_intl_1.defineMessage({
                id: \\"header2\\",
                defaultMessage: \\"Hello World!\\"
            })));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message.",
      "id": "foo.bar.baz",
    },
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message",
      "id": "header",
    },
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message",
      "id": "header2",
    },
  ],
}
`;

exports[`emit asserts for nested 1`] = `
Object {
  "code": "intl.formatMessage({
    id: \\"HELLO.undefined.13.undefined\\",
    defaultMessage: \\"layer1 {name}\\"
}, {
    name: intl.formatMessage({
        id: \\"HELLO.undefined.6.undefined\\",
        defaultMessage: \\"layer2\\"
    }),
});
",
  "msgs": Array [
    Object {
      "defaultMessage": "layer1 {name}",
      "id": "HELLO.undefined.13.undefined",
    },
    Object {
      "defaultMessage": "layer2",
      "id": "HELLO.undefined.6.undefined",
    },
  ],
}
`;

exports[`emit asserts for noImport 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function foo() {
    props.intl.formatMessage({
        id: \\"cRtnZ_\\",
        defaultMessage: \\"props {intl}\\"
    }, { bar: 'bar' });
    this.props.intl.formatMessage({
        id: \\"47JwLE\\",
        defaultMessage: \\"this props {intl}\\"
    }, { bar: 'bar' });
    return intl.formatMessage({
        id: \\"9XDP7L\\",
        defaultMessage: \\"foo {bar}\\"
    }, { bar: 'bar' });
}
exports.foo = foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "props {intl}",
      "description": "bar",
      "id": "cRtnZ_",
    },
    Object {
      "defaultMessage": "this props {intl}",
      "description": "bar",
      "id": "47JwLE",
    },
    Object {
      "defaultMessage": "foo {bar}",
      "description": "bar",
      "id": "9XDP7L",
    },
  ],
}
`;

exports[`emit asserts for overrideIdFn 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const msgs = react_intl_1.defineMessages({
    header: {
        id: \\"foo.bar.baz\\",
        defaultMessage: \\"Hello World!\\"
    },
    content: {
        id: \\"foo.bar.biff\\",
        defaultMessage: \\"Hello Nurse!\\"
    }
});
react_intl_1.defineMessage({
    id: \\"HELLO.undefined.13.string\\",
    defaultMessage: \\"defineMessage\\"
});
class Foo extends react_1.Component {
    render() {
        const { intl } = this.props;
        const { formatMessage } = intl;
        this.props.intl.formatMessage({
            id: \\"HELLO.undefined.5.string\\",
            defaultMessage: \\"no-id\\"
        });
        intl.formatMessage({
            id: \\"HELLO.undefined.18.string\\",
            defaultMessage: \\"intl.formatMessage\\"
        });
        formatMessage({
            id: \\"HELLO.undefined.13.string\\",
            defaultMessage: \\"formatMessage\\"
        });
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(\\"h1\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, Object.assign({}, msgs.header))),
            react_1.default.createElement(\\"p\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, Object.assign({}, msgs.content))),
            react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo.bar.zoo', defaultMessage: 'Hello World! {abc}', values: { abc: 2 } }),
            react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'HELLO.undefined.18.undefined', defaultMessage: 'Hello World! {abc}', values: { abc: 2 } })));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message",
      "id": "foo.bar.baz",
    },
    Object {
      "defaultMessage": "Hello Nurse!",
      "id": "foo.bar.biff",
    },
    Object {
      "defaultMessage": "defineMessage",
      "description": "foo",
      "id": "HELLO.undefined.13.string",
    },
    Object {
      "defaultMessage": "no-id",
      "description": "no-id",
      "id": "HELLO.undefined.5.string",
    },
    Object {
      "defaultMessage": "intl.formatMessage",
      "description": "no-id",
      "id": "HELLO.undefined.18.string",
    },
    Object {
      "defaultMessage": "formatMessage",
      "description": "no-id",
      "id": "HELLO.undefined.13.string",
    },
    Object {
      "defaultMessage": "Hello World! {abc}",
      "id": "foo.bar.zoo",
    },
    Object {
      "defaultMessage": "Hello World! {abc}",
      "id": "HELLO.undefined.18.undefined",
    },
  ],
}
`;

exports[`emit asserts for removeDefaultMessage 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const messages = react_intl_1.defineMessages({
    foo: {
        id: \\"greeting-user\\"
    }
});
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'greeting-world' }));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello, {name}",
      "description": "Greeting the user",
      "id": "greeting-user",
    },
    Object {
      "defaultMessage": "Hello World!",
      "description": "Greeting to the world",
      "id": "greeting-world",
    },
  ],
}
`;

exports[`emit asserts for removeDescription 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const messages = react_intl_1.defineMessages({
    foo: {
        id: \\"greeting-user\\",
        defaultMessage: \\"Hello, {name}\\"
    }
});
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'greeting-world', defaultMessage: 'Hello World!' }));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello, {name}",
      "description": "Greeting the user",
      "id": "greeting-user",
    },
    Object {
      "defaultMessage": "Hello World!",
      "description": "Greeting to the world",
      "id": "greeting-world",
    },
  ],
}
`;
